beacon-chain/rpc/apimiddleware/custom_hooks.go

```diff
@@ -168,15 +170,82 @@ func wrapSignedContributionAndProofsArray(
 	return true, nil
 }
 
-// Posted graffiti needs to have length of 32 bytes, but client is allowed to send data of any length.
-func prepareGraffiti(endpoint apimiddleware.Endpoint, _ http.ResponseWriter, _ *http.Request) apimiddleware.ErrorJson {
+type phase0PublishBlockRequestJson struct {
+	Phase0Block *beaconBlockJson `json:"phase0_block"`
+	Signature   string           `json:"signature" hex:"true"`
+}
+
+type altairPublishBlockRequestJson struct {
+	AltairBlock *beaconBlockAltairJson `json:"altair_block"`
+	Signature   string                 `json:"signature" hex:"true"`
+}
+
+// setInitialPublishBlockPostRequest is triggered before we deserialize the request JSON into a struct.
+// We don't know which version of the block got posted, but we can determine it from the slot.
+// We know that both Phase 0 and Altair blocks have a Message field with a Slot field,
+// so we deserialize the request into a struct s, which has the right fields, to obtain the slot.
+// Once we know the slot, we can determine what the PostRequest field of the endpoint should be, and we set it appropriately.
+func setInitialPublishBlockPostRequest(endpoint *apimiddleware.Endpoint,
+	_ http.ResponseWriter,
+	req *http.Request,
+) (apimiddleware.RunDefault, apimiddleware.ErrorJson) {
+	s := struct {
+		Message struct {
+			Slot string
+		}
+	}{}
+
+	buf, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		return false, apimiddleware.InternalServerErrorWithMessage(err, "could not read body")
+	}
+	if err := json.Unmarshal(buf, &s); err != nil {
+		return false, apimiddleware.InternalServerErrorWithMessage(err, "could not read slot from body")
+	}
+	slot, err := strconv.ParseUint(s.Message.Slot, 10, 64)
+	if err != nil {
+		return false, apimiddleware.InternalServerErrorWithMessage(err, "slot is not an unsigned integer")
+	}
+	if slot < uint64(params.BeaconConfig().AltairForkEpoch) {
```

Comment:
We are comparing slot against epoch. Should we cover this with a unit test?

Reply:
I added missing tests for `setInitialPublishBlockPostRequest`.